core/services/workflows/syncer/workflow_registry.go

```diff
@@ -2,39 +2,182 @@ package syncer
+	orm      WorkflowRegistryDS
+	wg       sync.WaitGroup
+	mu       sync.Mutex
 }
 
-func (w *WorkflowRegistry) Start(ctx context.Context) error {
-	return nil
+func (w *workflowRegistry) Start(ctx context.Context) error {
+	return w.StartOnce(w.Name(), func() error {
+		w.wg.Add(1)
+		go func() {
+			defer w.wg.Done()
+			w.syncLoop()
+		}()
+
+		w.wg.Add(1)
+		go func() {
+			defer w.wg.Done()
+			w.updateStateLoop()
+		}()
+
+		return nil
+	})
 }
 
-func (w *WorkflowRegistry) Close() error {
-	return nil
+func (w *workflowRegistry) Sync(ctx context.Context, isInitialSync bool) error {
+	if w.syncer != nil {
+		return w.syncer.Sync(ctx, isInitialSync)
+	}
+	return errors.New("not implemented")
 }
 
-func (w *WorkflowRegistry) Ready() error {
+func (w *workflowRegistry) Close() error {
+	return w.StopOnce(w.Name(), func() error {
+		close(w.stopCh)
+		w.wg.Wait()
+		return nil
+	})
+}
+
+func (w *workflowRegistry) Ready() error {
 	return nil
 }
 
-func (w *WorkflowRegistry) HealthReport() map[string]error {
+func (w *workflowRegistry) HealthReport() map[string]error {
 	return nil
 }
 
-func (w *WorkflowRegistry) Name() string {
-	return "WorkflowRegistrySyncer"
+func (w *workflowRegistry) Name() string {
+	return name
 }
 
-func (w *WorkflowRegistry) SecretsFor(workflowOwner, workflowName string) (map[string]string, error) {
+func (w *workflowRegistry) SecretsFor(workflowOwner, workflowName string) (map[string]string, error) {
 	// TODO: actually get this from the right place.
 	return map[string]string{}, nil
 }
 
-func NewWorkflowRegistry() *WorkflowRegistry {
-	return &WorkflowRegistry{}
+var (
+	defaultTickInterval = 12 * time.Second
+)
+
+func NewWorkflowRegistry(
+	lggr logger.Logger,
+	orm WorkflowRegistryDS,
+	opts ...func(*workflowRegistry),
+) *workflowRegistry {
+	wr := &workflowRegistry{
+		stopCh:   make(services.StopChan),
+		updateCh: make(chan any),
+		errCh:    make(chan error, 1),
+		lggr:     lggr.Named(name),
+		orm:      orm,
+	}
+	for _, opt := range opts {
+		opt(wr)
+	}
+	return wr
+}
+
+func (w *workflowRegistry) syncLoop() {
+	ctx, cancel := w.stopCh.NewCtx()
+	defer cancel()
+
+	ticker := w.getTicker(ctx)
+
+	// Sync for a first time outside the loop; this means we'll check the cached state
+	// or start a remote sync immediately once spinning up syncLoop, as by default a ticker will
+	// fire for the first time at T+N, where N is the interval.
+	//
+	// TODO(mstreet3): metrics
+	w.lggr.Debug("starting initial sync with remote registry")
+	err := w.Sync(ctx, true)
+	if err != nil {
+		w.lggr.Errorw("failed to sync with remote registry", "error", err)
+		w.sinkError(err)
+	}
+
+	for {
+		select {
+		case <-w.stopCh:
+			return
+		case <-ticker:
+			w.lggr.Debug("starting regular sync with the remote registry")
+			err := w.Sync(ctx, false)
+			if err != nil {
+				w.lggr.Errorw("failed to sync with remote registry", "error", err)
+				w.sinkError(err)
+			}
+		}
+	}
+}
+
+func (w *workflowRegistry) updateStateLoop() {
```

Comment:
I'd recommend using a stateless syncer initially to reduce complexity.

Reply:
Removed. Was scratch code